src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -195,29 +206,88 @@ ValidatorList::load(
 }
 
 boost::filesystem::path
-ValidatorList::GetCacheFileName(PublicKey const& pubKey)
+ValidatorList::getCacheFileName(PublicKey const& pubKey) const
 {
     return dataPath_ / (filePrefix_ + strHex(pubKey));
 }
 
+// static
+Json::Value
+ValidatorList::buildFileData(
+    std::string const& pubKey,
+    ValidatorList::PublisherListCollection const& pubCollection,
+    boost::optional<std::uint32_t> forceVersion /* = {} */)
+{
+    Json::Value value(Json::objectValue);
+
+    auto const effectiveVersion = forceVersion
+        ? *forceVersion
+        : (pubCollection.remaining.empty() ? pubCollection.rawVersion : 2);
+
+    value[jss::manifest] = pubCollection.rawManifest;
+    value[jss::version] = effectiveVersion;
+    value[jss::public_key] = pubKey;
+
+    switch (effectiveVersion)
+    {
+        case 1: {
+            auto const& current = pubCollection.current;
+            value[jss::blob] = current.rawBlob;
+            value[jss::signature] = current.rawSignature;
+            // This is only possible if "downgrading" a v2 UNL to v1, for
+            // example for the /vl/ endpoint.
+            if (current.rawManifest &&
+                *current.rawManifest != pubCollection.rawManifest)
+                value[jss::manifest] = *current.rawManifest;
+            break;
+        }
+        case 2: {
+            Json::Value blobs(Json::arrayValue);
+
+            auto add = [&blobs, &outerManifest = pubCollection.rawManifest](
+                           PublisherList const& pubList) {
+                auto& blob = blobs.append(Json::objectValue);
+                blob[jss::blob] = pubList.rawBlob;
+                blob[jss::signature] = pubList.rawSignature;
+                if (pubList.rawManifest &&
+                    *pubList.rawManifest != outerManifest)
+                    blob[jss::manifest] = *pubList.rawManifest;
+            };
+
+            add(pubCollection.current);
+            for (auto const& [_, pending] : pubCollection.remaining)
+            {
+                (void)_;
+                add(pending);
+            }
+
+            value[jss::blobs_v2] = std::move(blobs);
+            break;
+        }
+        default:
+            value = Json::nullValue;
```

Comment:
Am I correct in assuming that we should never hit this case? Maybe its worth logging.

Reply:
We can hit this case based on user input. The `/vl/` endpoint will parse a number and pass it in.
```
$ curl -D - -k https://localhost:51235/vl/999/ED2677ABFFD1B33AC6FBC3062B71F1E8397C1505E1C42C64D11AD1B28FF73F4734
HTTP/1.1 400 Bad Request
Server: rippled-1.7.0-b3+DEBUG
Content-Type: application/json
Connection: close
Transfer-Encoding: chunked

null
```
The 400 result code indicates that the public key is valid, but the request is bad - in this case, because it requested version 999. If the public key is not valid, you get a 404.

Is it worth logging? Eh, couldn't hurt, but I think only at trace level.